connection_instrumented.go

```diff
@@ -56,6 +59,9 @@ func instrumentDriver(deets *ConnectionDetails, defaultDriverName string) (drive
 		newDriverName = instrumentedDriverName + "-" + nameSQLite3
 	}
 
+	sqlDriverLock.Lock()
```

Comment:
If the issue is "called twice" rather than "called in two places or called at the same time", I am not sure if this lock will help the situation since there is no `return` between this lock and the calling of `sql.Register()` so the function will be finally called, just with a locked delay, once the mutex was unlocked and the call will still be the second call. If the function should not be called twice, we should prevent the call itself instead make it be called later.

Reply:
Oh, I just quickly checked the code and it seems like what you did is coupling calls of `sql.Drivers()` and `sql.Register()` together, so once a driver was registered by `Register()` then the driver name will be included in the return of `Drivers()` and the boolean `found` prevent the second call. I think it could be ok.
Please let me take a look at the code some more to find...  if we can add a test case for this. I agree it could be hard to find a test method though. Or... if we could try to make it simpler... :-)

Reply:
Here is the test to reliably reproduce the issue: https://github.com/zepatrik/pop/commit/8aac4d65d412192ccaeb362d3fcfbc70a483e2bb
It produces the output:
```
go test -run Test_InstrumentationRace -v .
[POP] 2022/05/30 13:05:12 info - Skipping integration tests because SODA_DIALECT is blank or unset
=== RUN   Test_InstrumentationRace
[POP] 2022/05/30 13:05:12 warn - SQL driver instrumentation was enabled but no options have been passed to `ConnectionDetails.InstrumentedDriverOptions`. Instrumentation will therefore not result in any output.
[POP] 2022/05/30 13:05:12 warn - SQL driver instrumentation was enabled but no options have been passed to `ConnectionDetails.InstrumentedDriverOptions`. Instrumentation will therefore not result in any output.
Found is false
Found is false
    connection_instrumented_test.go:146: 
                Error Trace:    connection_instrumented_test.go:146
                Error:          Received unexpected error:
                                recover returned: sql: Register called twice for driver instrumented-sql-driver-postgres
                Test:           Test_InstrumentationRace
--- FAIL: Test_InstrumentationRace (0.00s)
FAIL
FAIL    github.com/gobuffalo/pop/v6     0.011s
FAIL
```

However, this test will dead-lock when using the lock.

## The problem

1. `instrumentDriver` first determines the driver name (deterministic)
2. Because the `sql` package has no way to atomically determine whether a driver is already registered, we iterate over all drivers. The result is stored in the `found` variable.
3. If the driver was not found (and only then), the driver is registered.

However, during parallel calls of `NewConnection` we can have two goroutines that run step 2 in parallel. Both will determine that the driver is not registered yet (`found == false`) and will try to register it. One of them will do so successfully, however the second one will cause the mentioned panic.

## The fix

Adding the lock **before** the search & register part (before step 2) and releasing it after that part ensures that only ever one goroutine can be in this critical part of the function. Therefore, no goroutines will be able to have `found == false` at the same time, and therefore the panic will not happen ever again.

To come up with a test that reliably reproduces the problem, but also passes with the fix, proved to be impossible because we cannot reach the critical synchronization point with the fix, but need to reach it to reliably reproduce it.

Reply:
Edit: I figured out a test that works for both cases.